### PR TITLE
Feat/en 2648 processors integration

### PR DESCRIPTION
**Pull request by sasurobert**
integration of multiple processors and intermediate processors into transaction coordinator

### CODE REVIEW

File: process/block/preprocess/intermediateResults_test.go [hunk 1]
```diff
@@ -2,15 +2,18 @@ package preprocess
 
 import (
 	"bytes"
+	"github.com/stretchr/testify/assert"
```
Comment (ccorcoveanu) [inline process/block/preprocess/intermediateResults_test.go]: Imports are not sorted

File: process/coordinator/process.go [hunk 2]
```diff
@@ -431,6 +438,40 @@ func (tc *transactionCoordinator) CreateMbsAndProcessTransactionsFromMe(maxTxRem
 		}
 	}
 
+	interMBs := tc.processAddedInterimTransactions()
```
Comment (ccorcoveanu) [inline process/coordinator/process.go]: Why are we doing this two times? ( If we have to choose I like the one liner solution from line 446 better :) )

File: process/coordinator/process_test.go [hunk 3]
```diff
@@ -2,6 +2,10 @@ package coordinator
 
 import (
 	"bytes"
+	"github.com/ElrondNetwork/elrond-go/core"
```
Comment (ccorcoveanu) [inline process/coordinator/process_test.go]: Imports are not sorted

File: process/factory/containers/intermediateTransactionHandlersContainer.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+func (ppc *IntermediateTransactionHandlersContainer) Remove(key block.Type) {
+	ppc.objects.Del(uint8(key))
+}
+
+// Len returns the length of the added objects
+func (ppc *IntermediateTransactionHandlersContainer) Len() int {
+	return ppc.objects.Len()
+}
+
+func (ppc *IntermediateTransactionHandlersContainer) Keys() []block.Type {
```
Comment (ccorcoveanu) [inline process/factory/containers/intermediateTransactionHandlersContainer.go]: Comment missing on exported function

File: process/factory/containers/preProcessorsContainer.go [hunk 5]
```diff
@@ -86,3 +86,17 @@ func (ppc *PreProcessorsContainer) Remove(key block.Type) {
 func (ppc *PreProcessorsContainer) Len() int {
 	return ppc.objects.Len()
 }
+
+func (ppc *PreProcessorsContainer) Keys() []block.Type {
```
Comment (ccorcoveanu) [inline process/factory/containers/preProcessorsContainer.go]: There are repeating functions on different containers having `objects *hashmap.HashMap`. Can't we have a common implementation? Maybe use struct embedding to reuse the code

File: process/block/shardblock.go [hunk 6]
```diff
@@ -221,6 +221,11 @@ func (sp *shardProcessor) ProcessBlock(
 		return err
 	}
 
+	err = sp.txCoordinator.VerifyCreatedBlockTransactions(body)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/shardblock.go]: no. because the transactions are created at processing.

File: process/coordinator/process.go (hunk 2 above)
Comment (sasurobert) [reply process/coordinator/process.go]: done

File: process/factory/containers/intermediateTransactionHandlersContainer.go (hunk 4 above)
Comment (sasurobert) [reply process/factory/containers/intermediateTransactionHandlersContainer.go]: done

File: process/coordinator/process_test.go (hunk 3 above)
Comment (sasurobert) [reply process/coordinator/process_test.go]: done

File: process/block/preprocess/intermediateResults_test.go (hunk 1 above)
Comment (sasurobert) [reply process/block/preprocess/intermediateResults_test.go]: done

File: data/block/block.go [hunk 7]
```diff
@@ -26,14 +26,16 @@ type Body []*MiniBlock
...
+	// TxBlock identifies a miniblock holding transactions
 	TxBlock Type = 0
-	// StateBlock identifies a block holding account state
+	// StateBlock identifies a miniblock holding account state
 	StateBlock Type = 1
-	// PeerBlock identifies a block holding peer assignation
+	// PeerBlock identifies a miniblock holding peer assignation
 	PeerBlock Type = 2
-	// SmartContractResult identifies a block holding miniblocks
+	// SmartContractResult identifies a miniblock holding smartcontractresults
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply data/block/block.go]: done

File: process/factory/containers/preProcessorsContainer.go (hunk 5 above)
Comment (sasurobert) [reply process/factory/containers/preProcessorsContainer.go]: it is a container for a certain type, for a specific interface. One general would return interface{} and you would need type assertions for every get.

Comment (ccorcoveanu) [reply process/factory/containers/preProcessorsContainer.go]: `func (ppc *PreProcessorsContainer) Keys() []block.Type` returns `[]block.Type` just as the one in the `intermediateTransactionHandlersContainer.go`, maybe I am missing something :)

Comment (sasurobert) [reply process/factory/containers/preProcessorsContainer.go]: Keys return the same thing. but 60% of the public methods return different elements Preprocessor vs IntermediateProcessor. Get(key block.Type) (PreProcessor, error)

File: cmd/node/factory/structs.go [hunk 8]
```diff
@@ -1252,7 +1253,8 @@ func newShardBlockProcessorAndTracker(
 		return nil, nil, err
 	}
 
-	intermediateProcessor, err := preprocess.NewIntermediateResultsProcessor(core.Hasher, core.Marshalizer, shardCoordinator, state.AddressConverter)
+	intermediateProcessor, err := preprocess.NewIntermediateResultsProcessor(core.Hasher, core.Marshalizer,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: done.

File: integrationTests/multiShard/block/testInitializer.go [hunk 9]
```diff
@@ -257,15 +257,28 @@ func createNetNode(
 		&mock.SCProcessorMock{},
 	)
 
-	txCoordinator, _ := coordinator.NewTransactionCoordinator(
+	factory, _ := shard.NewPreProcessorsContainerFactory(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/block/testInitializer.go]: discussed.

File: process/block/preprocess/intermediateResults.go [hunk 10]
```diff
@@ -57,32 +62,40 @@ func NewIntermediateResultsProcessor(
...
 	miniBlocks := make([]*block.MiniBlock, irp.shardCoordinator.NumberOfShards())
 	for i := uint32(0); i < irp.shardCoordinator.NumberOfShards(); i++ {
 		miniBlocks[i] = &block.MiniBlock{}
 	}
 
 	irp.mutInterResultsForBlock.Lock()
-	defer irp.mutInterResultsForBlock.Unlock()
 
 	for key, value := range irp.interResultsForBlock {
 		if value.receiverShardID != irp.shardCoordinator.SelfId() {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/intermediateResults.go]: done.

**Pull request merged**